File: jl_src/day_14.jl
import ArgParse

# Build ArgParsing Table
s = ArgParse.ArgParseSettings()
@ArgParse.add_arg_table s begin
    "--input"
        help = "Filepath to input data"
        arg_type = String
        required = true
    "--verbose"
        help = "Add additional reporting"
        action = :store_true
end

# Parse arguments
parsed_args = ArgParse.parse_args(ARGS, s)

# Process files
function ProcessFile(filepath::String)::Tuple{String,Vector{Tuple{String,String}}}
    open(filepath) do f
        lines = readlines(f)
        rules = Vector{Tuple{String,String}}()

        init_formula = lines[1]

        for line in lines[3:end]
            split_line = split(line)
            left_val = split_line[1]
            right_val = split_line[3]
            push!(rules, (left_val, right_val))
        end
        return (init_formula, rules)
    end
end

function BuildRuleTable(in_rules::Vector{Tuple{String,String}})::Dict{String,Tuple{String,String}}
    rules = Dict{String,Tuple{String,String}}()
    for rule in in_rules
        left = rule[1]
        right = rule[2]
        left_pat = left[1]*right[1]
        right_pat = right[1]*left[2]
        rules[left] = (left_pat, right_pat)
    end
    rules
end

struct FormulaToTokens{T} end
function FormulaToTokens{T}(formula::String)::Dict{String,T} where T
    tokens = Dict{String,T}()
    for i in 1:(length(formula)-1)
        token = formula[i]*formula[i+1]
        tokens[token] = get(tokens, token, 0)+1
    end
    tokens
end

function ExpandFormula(formula::Dict{String,T}, rules::Dict{String,Tuple{String,String}})::Dict{String,T} where T
    new_formula = Dict{String,T}()
    for (key, val) in formula
        new_keys = rules[key]
        new_formula[new_keys[1]] = get(new_formula, new_keys[1], 0)+val
        new_formula[new_keys[2]] = get(new_formula, new_keys[2], 0)+val
    end
    new_formula
end

function CharCount(formula_tokens::Dict{String,T}, first_char::Char)::Dict{Char,T} where T
    char_count = Dict{Char,T}()
    char_count[first_char] = get(char_count, first_char, 0)+1
    for (token, num) in formula_tokens
       the_char = token[2]
       char_count[the_char] = get(char_count, the_char, 0)+num
    end
    char_count
end

function ComputeDiff(char_count::Dict{Char,T})::T where T
    min_v = minimum(values(char_count))
    max_v = maximum(values(char_count))
    return max_v-min_v
end

function SolveProblem(formula_tokens::Dict{String,T}, rules::Dict{String,Tuple{String,String}}) where T
    counts = Vector{UInt}()
    push!(counts, ComputeDiff(CharCount(formula_tokens, first_char)))

    num_steps = 40

    for _ in 1:num_steps
        formula_tokens = ExpandFormula(formula_tokens, rules)
        push!(counts, ComputeDiff(CharCount(formula_tokens, first_char)))
    end

    println("Day 14 problem 1: ", counts[11]);
    println("Day 14 problem 2: ", counts[41]);
end

# Get input filepath
input_filepath = parsed_args["input"]
verbose = parsed_args["verbose"]

init_formula, rules = ProcessFile(input_filepath)
first_char = init_formula[1]

rules = BuildRuleTable(rules)
formula_tokens = FormulaToTokens{UInt}(init_formula)

SolveProblem(formula_tokens, rules)

File: rust_src/src/day_14.rs
extern crate argparse;
extern crate regex;

use std::io::{self, BufRead};
use std::fs::File;
use std::path::Path;
use std::collections::HashMap;
use std::ops::{Add,Sub};
use std::fmt::Debug;
use argparse::{ArgumentParser, StoreTrue, Store};
use regex::Regex;

// Define readlines function discussed here:
// https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn process_input<P>(filename: P) -> (String, Vec<(String,String)>)
where
    P: AsRef<Path>,
{
    // Reading the file
    if let Ok(mut lines) = read_lines(filename) {
        // Consumes the iterator, returns an (Optional) String
        let mut rules: Vec<(String,String)> = Vec::new();
        let init_formula = lines.next().unwrap().unwrap();
        lines.next();
        let rule_re = Regex::new(r"([A-Z]*) -> ([A-Z]*)").unwrap();
        for line in lines {
            if let Ok(ip) = line {
                match rule_re.captures(&ip) {
                    Some(cap) => {
                        let left = cap.get(1).unwrap().as_str().to_string();
                        let right = cap.get(2).unwrap().as_str().to_string();
                        rules.push((left, right));
                    },
                    None => {
                    },
                }
            }
        }
        (init_formula, rules)
    } else {
		panic!("Problem opening file!");
	}
}

fn build_rule_table(in_rules: &Vec<(String,String)>) -> HashMap<String,(String,String)>
{
    let mut rules: HashMap<String,(String,String)> = HashMap::new();
    for (left, right) in in_rules {
        let left = left.chars().collect::<Vec::<char>>();
        let right = right.chars().collect::<Vec::<char>>();
        let left_pat: String = [left[0],right[0]].iter().collect();
        let right_pat: String = [right[0],left[1]].iter().collect();
        rules.insert(left.iter().collect(), (left_pat, right_pat));
    }
    rules
}

fn formula_to_tokens<T>(formula: &String) -> HashMap<String, T>
where
    T: Add+Add<Output=T>,
    T: num::NumCast,
    T: Copy,
{
    let mut tokens: HashMap<String, T> = HashMap::new();
    let formula_chars: Vec<char> = formula.chars().collect();
    for i in 0..formula_chars.len()-1 {
        let token: String = [formula_chars[i],formula_chars[i+1]].iter().collect();
        let val = tokens.entry(token).or_insert(num::cast::<_,T>(0).unwrap());
        *val = *val + num::cast::<_,T>(1).unwrap();
    }
    tokens
}

fn expand_formula<T>(formula: &HashMap<String,T>, rules: &HashMap<String,(String,String)>) -> HashMap<String,T>
where
    T: Add+Add<Output=T>,
    T: num::NumCast,
    T: Copy,
    T: Debug,
{
    //println!("======= new expand formula: {:?}", formula);
    let mut new_formula: HashMap<String,T> = HashMap::new();
    for (key, val) in formula.iter() {
        //println!("Transforming key: {}", key);
        let new_keys = rules.get(key).unwrap();
        //println!("new keys: {:?}", new_keys);

        let old_val = new_formula.entry(new_keys.0.clone()).or_insert(num::cast::<_,T>(0).unwrap());
        *old_val = *old_val+*val;
        let old_val = new_formula.entry(new_keys.1.clone()).or_insert(num::cast::<_,T>(0).unwrap());
        *old_val = *old_val+*val;
        //println!("Updated new formula: {:?}", new_formula);
    }
    //println!("new formula: {:?}", new_formula);
    new_formula
}

fn char_count<T>(formula_tokens: &HashMap<String,T>, first_char: char) -> HashMap<char,T>
where
    T: Add+Add<Output=T>,
    T: num::NumCast,
    T: Copy,
{
    let mut char_count: HashMap<char,T> = HashMap::new();
    let old_val = char_count.entry(first_char).or_insert(num::cast::<_,T>(0).unwrap());
    *old_val = *old_val+num::cast::<_,T>(1).unwrap();
    for (token, num) in formula_tokens.iter() {
       let the_char = token.chars().nth(1).unwrap();
       let old_val = char_count.entry(the_char).or_insert(num::cast::<_,T>(0).unwrap());
       *old_val = *old_val+(*num);
    }
    char_count
}

fn compute_diff<T>(char_count: &HashMap<char,T>) -> T
where
    T: Add+Add<Output=T>,
    T: Sub+Sub<Output=T>,
    T: num::NumCast,
    T: Copy,
    T: Ord,
{
    let min_v = char_count.values().min().unwrap();
    let max_v = char_count.values().max().unwrap();
    return *max_v-*min_v;
}

fn main() {
    // Argument Parsing
    let mut verbose = false;
    let mut input = String::new();
    {
        let mut ap = ArgumentParser::new();
        ap.set_description("Day 14 AOC 2021");
        ap.refer(&mut input)
            .add_option(&["-i", "--input"], Store,
            "Filepath to input data");
        ap.refer(&mut verbose)
            .add_option(&["-v", "--verbose"], StoreTrue,
            "Add additional reporting");
        ap.parse_args_or_exit();
    }

    let (init_formula, rules) = process_input(input);
    let first_char = init_formula.chars().nth(0).unwrap();

    let rules = build_rule_table(&rules);
    let mut formula_tokens = formula_to_tokens::<usize>(&init_formula);

    let mut counts: Vec<usize> = Vec::new();
    counts.push(compute_diff::<usize>(&char_count::<usize>(&formula_tokens, first_char)));

    let num_steps: usize = 40;
    for _ in 0..num_steps {
        formula_tokens = expand_formula::<usize>(&formula_tokens, &rules);
        counts.push(compute_diff::<usize>(&char_count::<usize>(&formula_tokens, first_char)));
    }

    println!("Day 14 problem 1: {}", counts[10]);
    println!("Day 14 problem 2: {}", counts[40]);
}
